server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -271,6 +275,35 @@ public void onFailure(Exception e) {
         }
     }
 
+    static void prohibitAppendOnlyWritesInBackingIndices(DocWriteRequest<?> writeRequest, Metadata metadata) {
+        IndexAbstraction indexAbstraction = metadata.getIndicesLookup().get(writeRequest.index());
+        if (indexAbstraction == null) {
+            return;
+        }
+        if (indexAbstraction.getType() != IndexAbstraction.Type.CONCRETE_INDEX) {
+            return;
+        }
+        if (indexAbstraction.getParentDataStream() == null) {
+            return;
+        }
+
+        // At this point with write op is targeting a data stream directly,so
+        // checking if write op is append-only and if so fail.
+        // (Updates and deletes are allowed to target a backing index)
+
+        DocWriteRequest.OpType opType = writeRequest.opType();
+        // CREATE op_type is considered append-only and
+        // INDEX op_type is considered append-only when no if_primary_term and if_seq_no is specified.
+        // (the latter maybe an update, but at this stage we can't determine that. In order to determine
+        // that an engine level change is needed and for now this check is sufficient.)
+        if (opType == DocWriteRequest.OpType.CREATE ||
+            (opType == DocWriteRequest.OpType.INDEX && writeRequest.ifPrimaryTerm() == UNASSIGNED_PRIMARY_TERM &&
+                writeRequest.ifSeqNo() == UNASSIGNED_SEQ_NO)) {
+            throw new IllegalArgumentException("append-only write targeting backing indices is disallowed," +
+                "target corresponding data stream instead");
```

Comment:
Minor: added space in error message
```suggestion
            throw new IllegalArgumentException("append-only write targeting backing indices is disallowed, " +
                "target corresponding data stream instead");
```